File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/null_object/solution/method_invocation.cpp
// method_invocation.cpp

#include <cmath>
#include <string>
#include "dp_const_iterator.h"
#include "dp_iterator.h"
#include "dp_vector.h"
class expression;
#include "expression_visitor.h"
#include "method_invocation.h"

method_invocation::method_invocation(const std::string & a_name)
{
	set_name(a_name);
}

void method_invocation::accept(expression_visitor & a_expression_visitor)
{
	a_expression_visitor.visit(*this);
}

void method_invocation::set_name(const std::string & a_name)
{
	my_name = a_name;
}

const std::string & method_invocation::get_name() const
{
	return my_name;
}

void method_invocation::add_operand(expression * a_operand)
{
	my_operands.add(a_operand);
}

dp_const_iterator<expression *> * method_invocation::get_operands() const
{
	return my_operands.make_iterator();
}

dp_iterator<expression *> * method_invocation::get_operands()
{
	return my_operands.make_iterator();
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/null_object/lab/MethodInvocation.java
// MethodInvocation.java

public class MethodInvocation extends Nonterminal
{
	public MethodInvocation(String aName)
	{
		setName(aName);
		myOperands = new DpVector<Expression>();
	}

	public void accept(ExpressionVisitor aExpressionVisitor)
	{
		aExpressionVisitor.visit(this);
	}

	private void setName(String aName)
	{
		myName = aName;
	}

	public String getName()
	{
		return myName;
	}

	public void addOperand(Expression aOperand)
	{
		myOperands.add(aOperand);
	}

	public DpIterator<Expression> getOperands()
	{
		return myOperands.makeIterator();
	}

	private String myName;
	private DpVector<Expression> myOperands;
}
